src/parquet/bloom.cc

```diff
@@ -0,0 +1,215 @@
+  case MURMUR3_X64_128:
+    this->hashFunc = &MurmurHash3_x64_128;
+    break;
+  default:
+    throw parquet::ParquetException("Not supported hash strategy");
+  }
+}
+
+void Bloom::setMask(uint32_t key, uint32_t mask[8]) {
+  for (int i = 0; i < 8; ++i) {
+    mask[i] = key * SALT[i];
+  }
+
+  for (int i = 0; i < 8; ++i) {
+    mask[i] = mask[i] >> 27;
+  }
+
+  for (int i = 0; i < 8; ++i) {
+    mask[i] = 0x1U << mask[i];
+  }
+}
+
+uint32_t Bloom::optimalNumOfBits(uint32_t ndv, double fpp) {
+  DCHECK(fpp > 0.0 && fpp < 1.0);
+  const double M = -8.0 * ndv / log(1 - pow(fpp, 1.0/8));
+  const double MAX = Bloom::DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES << 3;
+  int num_bits = static_cast<uint32_t>(M);
+
+  // Handle overflow.
+  if (M > MAX || M < 0) {
+    num_bits = static_cast<uint32_t>(MAX);
+  }
+
+  // Get next power of 2 if bits is not power of 2.
+  if ((num_bits & (num_bits - 1)) != 0) {
+    num_bits = static_cast<uint32_t>(::arrow::BitUtil::NextPower2(num_bits));
+  }
+
+  // Minimum
+  if (num_bits < (Bloom::BYTES_PER_FILTER_BLOCK << 3)) {
+    num_bits = Bloom::BYTES_PER_FILTER_BLOCK << 3;
+  }
+
+  return num_bits;
+}
+
+void Bloom::addElement(uint64_t hash) {
+  uint32_t* const bitset32 = bitset.get();
+  const uint32_t bucketIndex = static_cast<uint32_t>(hash >> 32)
+      & (num_bytes / BYTES_PER_FILTER_BLOCK - 1);
+  uint32_t key = static_cast<uint32_t>(hash);
+
+  // Calculate mask for bucket.
+  uint32_t mask[8];
+  setMask(key, mask);
+
+  for (int i = 0; i < 8; i++) {
+    bitset32[bucketIndex * 8 + i] |= mask[i];
+  }
+}
+
+bool Bloom::contains(uint64_t hash) {
+  uint32_t * const bitset32 = bitset.get();
+  const uint32_t bucketIndex = static_cast<uint32_t>((hash >> 32)
+      & (num_bytes / BYTES_PER_FILTER_BLOCK - 1));
+  uint32_t key = static_cast<uint32_t>(hash);
+
+  // Calculate mask for bucket.
+  uint32_t mask[8];
+  setMask(key, mask);
+
+  for (int i = 0; i < 8; ++i) {
+    if (0 == (bitset32[8 * bucketIndex + i] & mask[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Bloom::find(uint64_t hash) {
+  return contains(hash);
+}
+
+void Bloom::insert(uint64_t hash) {
+  addElement(hash);
+}
+
+uint64_t Bloom::hash(int value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(int), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(int64_t value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void*>(&value), sizeof(int64_t), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+uint64_t Bloom::hash(float value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(float), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+
+uint64_t Bloom::hash(double value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(&value), sizeof(double), DEFAULT_SEED, &out);
+  return out[0];
+}
+
+
+uint64_t Bloom::hash(const Int96 *value) {
+  uint64_t out[2];
+  (*hashFunc)(reinterpret_cast<void *>(const_cast<uint32_t*>(value->value)),
```

Comment:
This `const_cast` is not necessarily standards-compliant. I expect the type signature of the hash function is a problem: it should not be modifying the data passed into it, so it should be OK with a `const` pointer.

Reply:
yes, the const_cast is not necessarily and the signature is right also, I need to change reinterpret_cast<void*> to reinterpret_cast<const void *>